sim/shaman/fire_elemental_pet.go

```diff
@@ -0,0 +1,147 @@
+)
+
+type FireElemental struct {
+	core.Pet
+
+	FireBlast *core.Spell
+	FireNova  *core.Spell
+
+	FireShieldAura *core.Aura
+
+	shamanOwner *Shaman
+}
+
+func (shaman *Shaman) NewFireElemental() *FireElemental {
+	fireElemental := &FireElemental{
+		Pet: core.NewPet(
+			"Greater Fire Elemental",
+			&shaman.Character,
+			fireElementalPetBaseStats,
+			shaman.fireElementalStatInheritance(),
+			false,
+			true,
+		),
+		shamanOwner: shaman,
+	}
+	fireElemental.EnableManaBar()
+	fireElemental.EnableAutoAttacks(fireElemental, core.AutoAttackOptions{
+		MainHand: core.Weapon{
+			BaseDamageMin:  1,  // Estimated from base AP
+			BaseDamageMax:  24, // Estimated from base AP
+			SwingSpeed:     2,
+			SwingDuration:  time.Second * 2,
+			CritMultiplier: 2, // Pretty sure this is right.
+			SpellSchool:    core.SpellSchoolFire,
+		},
+		AutoSwingMelee: true,
+	})
+
+	fireElemental.OnPetEnable = fireElemental.enable
+	fireElemental.OnPetDisable = fireElemental.disable
+
+	shaman.AddPet(fireElemental)
+
+	return fireElemental
+}
+
+func (fireElemental *FireElemental) enable(sim *core.Simulation) {
+	fireElemental.FireShieldAura.Activate(sim)
+}
+
+func (fireElemental *FireElemental) disable(sim *core.Simulation) {
+	fireElemental.FireShieldAura.Deactivate(sim)
+}
+
+func (fireElemental *FireElemental) GetPet() *core.Pet {
+	return &fireElemental.Pet
+}
+
+func (fireElemental *FireElemental) Initialize() {
+	fireElemental.registerFireBlast()
+	fireElemental.registerFireNova()
+	fireElemental.registerFireShieldAura()
+}
+
+func (fireElemental *FireElemental) Reset(sim *core.Simulation) {
+}
+
+func (fireElemental *FireElemental) OnGCDReady(sim *core.Simulation) {
+	target := fireElemental.CurrentTarget
+
+	//Check for mana issues first
+	if fireElemental.CurrentMana() < fireElemental.FireNova.CurCast.Cost {
+		fireElemental.WaitForMana(sim, fireElemental.FireNova.CurCast.Cost)
+		return
+	}
+
+	//If no CD's are available on this GCD lets wait for the next spell off CD
+	if !fireElemental.FireBlast.IsReady(sim) && !fireElemental.FireNova.IsReady(sim) {
+		waitingOnCD := core.MinDuration(fireElemental.FireBlast.TimeToReady(sim), fireElemental.FireNova.TimeToReady(sim))
+		fireElemental.WaitUntil(sim, sim.CurrentTime+waitingOnCD)
+		return
+	}
+
+	numberCasts := fireElemental.FireBlast.SpellMetrics[0].Casts
+	if numberCasts < nFireBlastCasts && fireElemental.FireBlast.IsReady(sim) {
+		fireElemental.FireBlast.Cast(sim, target)
+		return
+	}
+
+	numberCasts = fireElemental.FireNova.SpellMetrics[0].Casts
+	if numberCasts < nFireNovaCasts && fireElemental.FireNova.IsReady(sim) {
+		fireElemental.FireNova.Cast(sim, target)
+		return
+	}
+
+}
+
+var fireElementalPetBaseStats = stats.Stats{
+	stats.Mana:        1789,
+	stats.Health:      994,
+	stats.Intellect:   147,
+	stats.Stamina:     327,
+	stats.SpellPower:  995,  //Estimated
+	stats.AttackPower: 1369, //Estimated
+
+	// TODO : No idea what his crit is at, he does not seem to gain any crit from owner.
```

Comment:
Added a row in my spreadsheet for Fire Shield crit testing; he seems to gain spell crit from *his own* intellect at a rate of about 1% crit per 212 intellect, and has a base spell crit of 2.61%.

I don't know whether the sim adds spell crit from int here, or if this is just the base and the scaling is added elsewhere.

He likely gains melee crit from his own agility as well, but there are only a few ways to increase it; Blessing of Kings and Scroll of Agility. Intellect is more relevant as he gains it directly from the player.